ring-core/src/ring/middleware/multipart_params.clj

```diff
@@ -144,32 +149,56 @@
-                       specified, uses the encoding specified in a part named
-                       \"_charset_\", or the content type for each part, or
-                       request character encoding if the part has no encoding,
-                       or \"UTF-8\" if no request character encoding is set.
-
-  :fallback-encoding - specifies the character encoding used in parsing if a
-                       part of the request does not specify encoding in its
-                       content type or no part named \"_charset_\" is present.
-                       Has no effect if :encoding is also set.
-
-  :store             - a function that stores a file upload. The function
-                       should expect a map with :filename, :content-type and
-                       :stream keys, and its return value will be used as the
-                       value for the parameter in the multipart parameter map.
-                       The default storage function is the temp-file-store.
-
-  :progress-fn       - a function that gets called during uploads. The
-                       function should expect four parameters: request,
-                       bytes-read, content-length, and item-count."
+  :encoding
+  - Character encoding to use for multipart parsing.
+    Overrides the encoding specified in the request. If not
+    specified, uses the encoding specified in a part named
+    \"_charset_\", or the content type for each part, or
+    request character encoding if the part has no encoding,
+    or \"UTF-8\" if no request character encoding is set.
+
+  :fallback-encoding
+  - Specifies the character encoding used in parsing if a
+    part of the request does not specify encoding in its
+    content type or no part named \"_charset_\" is present.
+    Has no effect if :encoding is also set.
+
+  :store
+  - A function that stores a file upload. The function
+    should expect a map with :filename, :content-type and
+    :stream keys, and its return value will be used as the
+    value for the parameter in the multipart parameter map.
+    The default storage function is the temp-file-store.
+
+  :progress-fn
+  - A function that gets called during uploads. The
+    function should expect four parameters: request,
+    bytes-read, content-length, and item-count.
+
+  :invalid-filename-handler
+  - A function that gets called when the file being uploaded
+    has an invalid name. The function should expect two
+    parameters: request and an exception of type
+    InvalidFileNameException. It should return a ring response."
   ([handler]
    (wrap-multipart-params handler {}))
   ([handler options]
    (fn
      ([request]
-      (handler (multipart-params-request request options)))
+      (let [req-or-ex (try
+                        (multipart-params-request request options)
+                        (catch Exception ex ex))
+            invalid-filename-handler
```

Comment:
This `let` binding can be moved out of the inner function so we don't need to perform a key lookup each request.